File: Communique/internal/protocol/library.go
package protocol

import (
	"regexp"
	"strings"

	"orc/cq/internal/fault"
)

// The library: the repository as something to read from a browser.
//
// Two views of the same tree. Documents are what Dock sees — markdown carrying
// § sections — and files are what is on disk, with Anno's annotations attached
// where a file has them. Both carry their whole text, because the two sides
// never meet: the server cannot ask the agent for a file, so anything the site
// can show is something a sync already carried.
//
// That is affordable and was measured before it was designed: the documentation
// is tens of kilobytes and the source is a couple of megabytes, against a
// snapshot limit of 32. Folding in the interface is about what a reader wants to
// look at, not about what would fit.

// Library size limits. A repository that exceeds them is reported as truncated
// rather than silently cut: a reader who cannot see a file must be able to tell
// that from a file that does not exist.
const (
	// MaxLibraryBytes bounds the whole library within one snapshot.
	MaxLibraryBytes = 24 << 20
	// MaxFileBytes bounds one file. A generated file of a million lines is not
	// something anybody browses, and holding it costs everything else its place.
	MaxFileBytes = 1 << 20
	// MaxLibraryFiles bounds how many files are carried at all.
	MaxLibraryFiles = 8192
	// MaxPathRunes bounds a path.
	MaxPathRunes = 1024
)

// Section is one § section of a document, as Dock reports it.
type Section struct {
	Number string `json:"number"`
	Name   string `json:"name"`
	Depth  int    `json:"depth"`
	Start  int    `json:"start"`
	End    int    `json:"end"`
	Lines  int    `json:"lines"`
	Out    int    `json:"out"`
	// In is how many sections link here, absent when it was not counted.
	In *int `json:"in,omitempty"`
}

// Validate checks the section is addressable and its span is coherent.
func (s Section) Validate() error {
	if s.Number == "" {
		return fault.Field("Section", "number", "section number is empty")
	}
	if err := checkText("Section", "number", s.Number, 64, false); err != nil {
		return err
	}
	if err := checkText("Section", "name", s.Name, MaxSubjectRunes, true); err != nil {
		return err
	}
	if s.Depth < 1 {
		return fault.Field("Section", "depth", "depth %d is not a heading level", s.Depth)
	}
	// A heading with nothing under it of its own is a real section, and Dock
	// reports its span as 0:0. Refusing that failed the whole sync the first time
	// somebody created a document from the browser and had not written the body
	// yet — one empty section costing an entire mirror.
	if s.Start == 0 && s.End == 0 {
		return nil
	}
	return checkSpan("Section", s.Start, s.End)
}

// Annotation is one Anno annotation, with the ones nested inside it.
type Annotation struct {
	Kind         string       `json:"kind"`
	Name         string       `json:"name"`
	Meta         []string     `json:"meta,omitempty"`
	Start        int          `json:"start"`
	End          int          `json:"end"`
	Lines        int          `json:"lines"`
	ContentStart int          `json:"content_start"`
	ContentEnd   int          `json:"content_end"`
	Children     []Annotation `json:"children,omitempty"`
}

// Validate checks the annotation and everything under it.
//
// Depth is bounded because the structure is recursive and arrives over a wire:
// without a limit a hostile or damaged snapshot could nest deeply enough to
// exhaust the stack while being validated.
func (a Annotation) Validate() error { return a.validate(0) }

// MaxAnnotationDepth is far above Anno's own three ranks, so it bounds damage
// without ever bounding a real file.
const MaxAnnotationDepth = 32

func (a Annotation) validate(depth int) error {
	if depth > MaxAnnotationDepth {
		return fault.Field("Annotation", "children", "nested more than %d deep", MaxAnnotationDepth)
	}
	switch a.Kind {
	case "section", "symbol", "part":
	default:
		return fault.Field("Annotation", "kind", "unknown kind %q", a.Kind)
	}
	if a.Name == "" {
		return fault.Field("Annotation", "name", "annotation name is empty")
	}
	if err := checkText("Annotation", "name", a.Name, MaxNameRunes, false); err != nil {
		return err
	}
	for _, m := range a.Meta {
		if err := checkText("Annotation", "meta", m, MaxSubjectRunes, true); err != nil {
			return err
		}
	}
	if err := checkSpan("Annotation", a.Start, a.End); err != nil {
		return err
	}
	for _, child := range a.Children {
		if err := child.validate(depth + 1); err != nil {
			return err
		}
	}
	return nil
}

// File is one file of the repository, with its text.
type File struct {
	Path  string `json:"path"`
	Lines int    `json:"lines"`
	Bytes int    `json:"bytes"`
	// Text is the file's contents, empty when it was too large to carry. Bytes
	// still reports its real size, so the interface can say what it is missing
	// rather than showing an empty file.
	Text string `json:"text,omitempty"`
	// Sections are Dock's view, present on documents.
	Sections []Section `json:"sections,omitempty"`
	// Annotations are Anno's view, present on files that carry them.
	Annotations []Annotation `json:"annotations,omitempty"`
	// Skipped says why the text is absent, empty when it is present.
	Skipped string `json:"skipped,omitempty"`
}

// Validate checks the file is addressable and internally consistent.
func (f File) Validate() error {
	if err := checkPath("File", "path", f.Path); err != nil {
		return err
	}
	if f.Bytes < 0 {
		return fault.Field("File", "bytes", "size %d is negative", f.Bytes)
	}
	if f.Text != "" && f.Skipped != "" {
		return fault.Field("File", "skipped", "a file with text also says why it has none")
	}
	if err := checkText("File", "text", f.Text, MaxFileBytes, true); err != nil {
		return err
	}
	for _, s := range f.Sections {
		if err := s.Validate(); err != nil {
			return err
		}
	}
	for _, a := range f.Annotations {
		if err := a.Validate(); err != nil {
			return err
		}
	}
	return nil
}

// Library is the readable repository.
type Library struct {
	// Root is what the paths are relative to, for display only.
	Root  string `json:"root"`
	Files []File `json:"files"`
	// Truncated says what was left out, empty when nothing was. A reader who
	// cannot find a file must be able to tell "too big to carry" from "not
	// there", and silence would make those the same.
	Truncated string `json:"truncated,omitempty"`
	// Notes are things that went wrong while collecting, in the reader's terms.
	//
	// A lens that could not be run is the case this exists for: without it, a
	// missing `dock` shows up as a documentation tab confidently reporting that
	// no document carries a section — a claim about the operator's files that is
	// not true, when the truth is that a tool is not installed.
	Notes []string `json:"notes,omitempty"`
}

// Validate checks the library and every file in it.
func (l Library) Validate() error {
	if err := checkText("Library", "root", l.Root, MaxPathRunes, true); err != nil {
		return err
	}
	for _, note := range l.Notes {
		if err := checkText("Library", "notes", note, MaxSubjectRunes, false); err != nil {
			return err
		}
	}
	if len(l.Files) > MaxLibraryFiles {
		return fault.Field("Library", "files", "%d files exceeds the limit of %d", len(l.Files), MaxLibraryFiles)
	}
	seen := make(map[string]bool, len(l.Files))
	for _, f := range l.Files {
		if err := f.Validate(); err != nil {
			return err
		}
		// One path, one file. A duplicate means the two ends disagree about
		// what a path addresses, and the interface would show whichever it
		// happened to reach first.
		if seen[f.Path] {
			return fault.Field("Library", "files", "%q appears twice", f.Path)
		}
		seen[f.Path] = true
	}
	return nil
}

// CheckPath is the path rule, for a caller outside this package.
//
// It is exported so the server can refuse a bad path when it arrives rather than
// only when the action it becomes is validated. One rule, in one place: a second
// copy in the handler would be a second opinion about what a path is.
func CheckPath(field, path string) error { return checkPath("request", field, path) }

// CheckText is the text rule, for the same reason. Empty is allowed: an empty
// file is a real file.
func CheckText(field, value string, max int) error {
	return checkText("request", field, value, max, true)
}

// checkSpan validates a one-based inclusive line range.
func checkSpan(where string, start, end int) error {
	if start < 1 {
		return fault.Field(where, "start", "line %d is not a line number", start)
	}
	if end < start {
		return fault.Field(where, "end", "range %d:%d ends before it begins", start, end)
	}
	return nil
}

// checkPath validates a repository-relative path.
//
// It refuses anything absolute or containing a parent step, because the
// interface turns these into links and the server looks files up by them: a
// path that climbs out of the tree is either a bug or an attempt to address
// something the library was never meant to carry.
func checkPath(where, field, path string) error {
	if path == "" {
		return fault.Field(where, field, "path is empty")
	}
	if err := checkText(where, field, path, MaxPathRunes, false); err != nil {
		return err
	}
	if strings.HasPrefix(path, "/") {
		return fault.Field(where, field, "path %q is absolute", path)
	}
	// A drive letter is absolute too, and `filepath.IsAbs` only knows that on
	// the platform that has drives — so the server would pass it along and only
	// the agent would notice.
	if len(path) >= 2 && path[1] == ':' && isLetter(path[0]) {
		return fault.Field(where, field, "path %q names a volume", path)
	}
	// A path here is separated by `/`, on every machine. A backslash is not a
	// separator to the server and is one to a Windows agent, so a path holding
	// one means two different things at the two ends of the wire — and the
	// splitting below, which is what refuses `..`, would be looking at the
	// wrong parts.
	if strings.ContainsRune(path, '\\') {
		return fault.Field(where, field, "path %q holds a backslash; paths are separated by / here, on every machine", path)
	}
	for _, part := range strings.Split(path, "/") {
		if part == ".." {
			return fault.Field(where, field, "path %q climbs out of the tree", path)
		}
	}
	return nil
}

func isLetter(b byte) bool {
	return (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z')
}

// hex64 is a SHA-256 rendered the way every other digest in this tree is.
var hex64 = regexp.MustCompile(`^[0-9a-f]{64}$`)

// checkDigest validates the precondition a library verb carries.
//
// It is the whole of what makes editing a mirror safe: a snapshot is minutes old
// by the time somebody acts on it, and an action that could not say what it
// expected to find would be an action that silently overwrites whatever arrived
// in between.
func checkDigest(where, field, digest string) error {
	if digest == "" {
		return fault.Field(where, field, "no digest of what was edited")
	}
	if !hex64.MatchString(digest) {
		return fault.Field(where, field, "digest %q is not 64 hex characters", digest)
	}
	return nil
}

File: Communique/internal/web/app/library.js
// The library: the repository, folded.
//
// Two tabs over one tree. Documents are the files Dock found § sections in, and
// code is everything else — the same data, filtered, because "what is written
// about this" and "what does it do" are different questions and answering both
// in one list would serve neither.
//
// Nothing here fetches. The tree arrives as structure without text, so every
// file and every section can be listed and folded before anything is read; a
// file's contents are asked for when a reader opens it, and once. That is what
// makes folding worth having rather than a way of hiding what was already
// downloaded.

import { h, ellipsis } from "./dom.js";
import { render } from "./markdown.js";
import { block } from "./highlight.js";
import { digest } from "./digest.js";

// plural counts a thing without the "1 files" that reads as a bug in a tool
// that is otherwise careful about its words.
export function plural(n, word) {
  return `${n} ${word}${n === 1 ? "" : "s"}`;
}

// tree turns the flat list of paths into nested directories.
//
// The server sends paths because a path is what addresses a file; the nesting is
// presentation, and rebuilding it here keeps the wire format flat and obvious.
export function tree(files) {
  const root = { dirs: new Map(), files: [] };
  for (const file of files) {
    const parts = file.path.split("/");
    let node = root;
    for (const dir of parts.slice(0, -1)) {
      if (!node.dirs.has(dir)) node.dirs.set(dir, { dirs: new Map(), files: [] });
      node = node.dirs.get(dir);
    }
    node.files.push(file);
  }
  return root;
}

// counts totals a directory, so a folded one still says how much is inside it.
export function counts(node) {
  let files = node.files.length;
  let lines = node.files.reduce((n, f) => n + (f.lines || 0), 0);
  for (const child of node.dirs.values()) {
    const inner = counts(child);
    files += inner.files;
    lines += inner.lines;
  }
  return { files, lines };
}

// emptyReason says why there is nothing to show, or nothing when there is.
//
// Three different things produce an empty tab and they have three different
// remedies, so they get three different sentences. Telling somebody that no
// machine mirrors a repository when the truth is that the request failed sends
// them to the wrong machine entirely.
export function emptyReason(state) {
  const lib = state.library;

  if (lib && lib.unreachable) {
    return [
      h("p", { class: "error" }, `the repository could not be fetched: ${lib.unreachable}`),
      h("p", { class: "muted" }, "an older cq on the server has no library to serve; rebuild and restart it"),
    ];
  }
  if (!lib) {
    return [h("p", { class: "muted" }, "no machine is mirroring a repository")];
  }
  if ((lib.files || []).length === 0) {
    return [
      h("p", { class: "muted" }, "nothing mirrored yet"),
      // The setting, and where it has to be set. It is read by `cq sync` on the
      // agent machine, which is not the machine serving this page — the single
      // most common way to set it and see no effect.
      h("p", { class: "muted" }, "set CQ_LIBRARY on the agent machine and sync; `cq status` there says whether it is set"),
    ];
  }
  return null;
}

// isDocument reports whether Dock found sections in a file. It is the split
// between the two tabs, and it is Dock's answer rather than a guess from the
// extension: a .md file with no § headings is not a document Dock can address.
export function isDocument(file) {
  return (file.sections || []).length > 0;
}

export function library(state, actions, { kind }) {
  const ctx = { state, actions };
  const empty = emptyReason(state);
  if (empty) return empty;

  const all = state.library.files || [];

  const files = all.filter((f) => (kind === "docs" ? isDocument(f) : !isDocument(f)));
  const notes = state.library.notes || [];

  if (files.length === 0) {
    // Why it is empty, when the collection knows why. Saying "no document
    // carries a § section" because `dock` was not installed is a confident
    // claim about the reader's files that happens to be false, and it sends
    // them to look in the wrong place.
    if (kind === "docs" && notes.length > 0) {
      return notes.map((n) => h("p", { class: "warn" }, n));
    }
    return [h("p", { class: "muted" }, kind === "docs"
      ? "no document carries a § section yet"
      : "every mirrored file is a document")];
  }

  const out = [];
  if (state.library.truncated) {
    // Said, not silently true: a reader who cannot find a file has to be able
    // to tell that from the file not existing.
    out.push(h("p", { class: "warn" }, state.library.truncated));
  }
  for (const note of notes) out.push(h("p", { class: "warn" }, note));
  const root = tree(files);
  out.push(...dirNodes(root, "", ctx, 0));
  // The root gets the same controls as any other directory. Without them the
  // one place you cannot make a file is the top of the tree, which is where a
  // new module starts.
  out.push(dirControls("", root, ctx));
  return out;
}

// fileKey identifies a file in the read-text cache.
//
// The machine is part of it because a path is only unique within one checkout.
// Two machines mirroring trees that both hold `Docs/Vision.md` would otherwise
// share one entry — showing one machine's text under the other's file, and
// handing an edit the wrong digest to claim it was made against.
export function fileKey(file) {
  return `${file.machine || ""}\u0000${file.path}`;
}

// dirNodes draws one directory: its subdirectories, then its files.
function dirNodes(node, prefix, ctx, depth) {
  const out = [];
  for (const [name, child] of [...node.dirs.entries()].sort()) {
    const path = prefix ? `${prefix}/${name}` : name;
    const inner = counts(child);
    out.push(fold({
      key: `dir:${path}`,
      ctx,
      depth,
      label: h("span", { class: "dir" }, `${name}/`),
      note: `${plural(inner.files, "file")} · ${plural(inner.lines, "line")}`,
      controls: () => dirControls(path, child, ctx),
      children: () => dirNodes(child, path, ctx, depth + 1),
    }));
  }
  for (const file of node.files.sort((a, b) => a.path.localeCompare(b.path))) {
    out.push(fileNode(file, ctx, depth));
  }
  return out;
}

// dirControls are what can be done to a directory.
//
// The machine comes from a file inside it, because a directory is not itself
// mirrored — it exists because things are in it, and an action has to name the
// machine that holds them.
function dirControls(path, node, ctx) {
  if (ctx.state.picked !== `dir:${path}`) return null;
  const machine = machineIn(node);

  return h("div", { class: "controls row" },
    h("button", { class: "quiet", onclick: () => ctx.actions.newFile(machine, path) }, "new file"),
    h("button", { class: "quiet", onclick: () => ctx.actions.newFolder(machine, path) }, "new folder"),
    // The root is the checkout itself. Removing it is not a folder operation,
    // and there is no `..` to be standing in afterwards.
    path === "" ? null : h("button", {
      class: "quiet danger",
      onclick: () => ctx.actions.removeFolder(machine, path, filesUnder(node), hollow(node)),
    }, "delete folder"),
  );
}

// filesUnder is every file the mirror is showing inside a directory, at any
// depth — the manifest a recursive removal carries, and the number the operator
// is asked to confirm.
//
// Skipped files are in it too. They are files the mirror knows about and named a
// reason for, so they are part of what the operator was shown, and leaving them
// out would make the agent refuse over a file that is on screen.
export function filesUnder(node) {
  const out = [];
  const walk = (n) => {
    for (const f of n.files) out.push(f.path);
    for (const child of n.dirs.values()) walk(child);
  };
  walk(node);
  return out.sort();
}

// hollow reports a directory with nothing in it at all — no files anywhere
// under it, and no subdirectories either.
//
// It picks the verb. A hollow one gets `rmdir`, which is the narrowest thing
// that does the job and gives the clearest refusal if it turns out not to be
// empty after all. Anything else gets the recursive verb and the manifest.
function hollow(node) {
  return node.files.length === 0 && node.dirs.size === 0;
}

// machineIn finds which machine holds a directory's contents.
function machineIn(node) {
  for (const f of node.files) if (f.machine) return f.machine;
  for (const child of node.dirs.values()) {
    const found = machineIn(child);
    if (found) return found;
  }
  return "";
}

function fileNode(file, ctx, depth) {
  const name = file.path.split("/").pop();
  const note = file.skipped
    ? file.skipped
    : `${plural(file.lines, "line")}${structureNote(file)}`;

  return fold({
    key: `file:${file.path}`,
    ctx,
    depth,
    label: h("span", { class: "file" }, name),
    note,
    // Opening a file is what asks for its text. Until then the row is drawn
    // from the tree alone, which is why the whole repository can be listed.
    onOpen: () => ctx.actions.openFile(file),
    // A file's controls need it to have been read: an edit and a delete each
    // carry the digest of what was on screen, and there is no digest of a file
    // nobody has opened. So unlike a folder's, these appear only once it is.
    controls: () => fileControls(file, ctx),
    children: () => fileBody(file, ctx, depth + 1),
  });
}

function structureNote(file) {
  const bits = [];
  const sections = (file.sections || []).length;
  const annotations = countAnnotations(file.annotations || []);
  if (sections > 0) bits.push(plural(sections, "section"));
  if (annotations > 0) bits.push(plural(annotations, "annotation"));
  return bits.length > 0 ? ` · ${bits.join(" · ")}` : "";
}

export function countAnnotations(nodes) {
  return nodes.reduce((n, a) => n + 1 + countAnnotations(a.children || []), 0);
}

// fileBody draws what is inside a file: its structure, and its text once read.
function fileBody(file, ctx, depth) {
  const state = ctx.state;
  const loaded = state.files ? state.files[fileKey(file)] : null;
  if (file.skipped) {
    return [h("p", { class: "muted" }, file.skipped)];
  }
  if (!loaded) return [h("p", { class: "muted" }, "reading…")];
  if (loaded.error) return [h("p", { class: "error" }, loaded.error)];

  const lines = (loaded.text || "").split("\n");
  const out = [];

  // Dock's sections and Anno's annotations are both foldable spans over the same
  // lines, so both are drawn the same way and a file may have either or both.
  for (const section of file.sections || []) {
    out.push(fold({
      key: `sec:${file.path}:${section.number}`,
      ctx,
      depth,
      label: h("span", { class: "sect" }, `§${section.number} ${section.name}`),
      note: plural(section.lines, "line"),
      children: () => [
        prose(file, slice(lines, section.start, section.end)),
        h("div", { class: "controls" },
          h("button", {
            class: "quiet",
            onclick: () => ctx.actions.editSpan(file, loaded.text || "",
              section.start, section.end, `§${section.number} ${section.name}`),
          }, "edit this section"),
        ),
      ],
    }));
  }
  for (const node of file.annotations || []) {
    out.push(annotationNode(file, node, lines, ctx, depth));
  }

  // A file with no structure is just its text, and hiding that behind another
  // fold would be a fold that always has exactly one thing in it.
  if (out.length === 0) {
    out.push(prose(file, loaded.text || ""));
  }
  return out;
}

// fileControls are what can be done to a whole file.
//
// They appear only once the file has been opened — the destructive one should be
// reached only by somebody who has looked at what they are deleting, and an edit
// carries the digest of text nobody has fetched yet otherwise.
function fileControls(file, ctx) {
  if (ctx.state.picked !== `file:${file.path}`) return null;
  const loaded = ctx.state.files ? ctx.state.files[fileKey(file)] : null;
  if (!loaded || loaded.error || file.skipped) return null;
  const text = loaded.text || "";
  return h("div", { class: "controls row" },
    h("button", {
      class: "quiet",
      onclick: () => ctx.actions.editFile(file, text),
    }, "edit"),
    h("button", {
      class: "quiet danger",
      onclick: () => ctx.actions.deleteFile(file),
    }, "delete"),
  );
}

function annotationNode(file, node, lines, ctx, depth) {
  return fold({
    key: `ann:${file.path}:${node.kind}:${node.name}:${node.start}`,
    ctx,
    depth,
    label: h("span", { class: "annot" }, `${node.kind} ${node.name}`),
    note: `${plural(node.lines, "line")}${node.meta && node.meta.length ? ` · ${node.meta.join(" ")}` : ""}`,
    children: () => {
      const inner = (node.children || []).map((c) => annotationNode(file, c, lines, ctx, depth + 1));
      // The annotation's own content, plus whatever is nested inside it. Both,
      // because a symbol's body is worth reading and so are its parts.
      return [
        block(slice(lines, node.content_start, node.content_end), file.path),
        h("div", { class: "controls" },
          h("button", {
            class: "quiet",
            onclick: () => ctx.actions.editSpan(file, wholeOf(ctx, file),
              node.content_start, node.content_end, `${node.kind} ${node.name}`),
          }, "edit this annotation"),
        ),
        ...inner,
      ];
    },
  });
}

// wholeOf is the file's whole text, which a span edit needs in order to splice
// its change back into it.
function wholeOf(ctx, file) {
  const loaded = ctx.state.files ? ctx.state.files[fileKey(file)] : null;
  return (loaded && loaded.text) || "";
}

function slice(lines, start, end) {
  // A span of 0:0 is a heading with nothing under it of its own, which Dock
  // reports for a document somebody has started and not written yet. Slicing it
  // as though it began at line 1 would show the reader the wrong lines.
  if (start === 0 && end === 0) return "";
  // Otherwise spans are one-based and inclusive, the way every Orc tool reports
  // them.
  return lines.slice(Math.max(start - 1, 0), end).join("\n");
}

// prose renders markdown for a document and highlighted source for anything
// else. Markdown has its own renderer and is not also highlighted: two of them
// over the same text would be one of them losing.
function prose(file, text) {
  if (file.path.endsWith(".md")) return h("div", { class: "body" }, render(text));
  return block(text, file.path);
}

// fold is one collapsible row.
//
// Openness lives in the application's state rather than in the DOM, so a redraw
// — which happens on every sync — does not collapse everything the reader had
// opened. That is the difference between a fold that works and one that fights.
function fold({ key, ctx, depth, label, note, children, controls, onOpen }) {
  const open = ctx.state.open ? ctx.state.open[key] : false;

  const head = h("button", {
    class: ctx.state.picked === key ? "fold picked" : "fold",
    "aria-expanded": open ? "true" : "false",
    "aria-current": ctx.state.picked === key ? "true" : null,
    // The depth is handed to CSS rather than turned into padding here: a phone
    // cannot afford a two-character indent five levels down a repository, and
    // that is a layout decision, which belongs in the stylesheet.
    //
    // An object, so it goes through the CSSOM. As a `style` attribute this never
    // reached the browser at all — the site's content policy has no
    // `unsafe-inline` — so every fold in the tree was drawn at depth zero. See
    // dom.js.
    style: { "--depth": depth },
    onclick: () => {
      if (!open && onOpen) onOpen();
      ctx.actions.toggle(key);
      // Picking is what puts this row's controls on screen. One row at a time,
      // because a tree with a button under every line is a tree nobody can read
      // — and the thing somebody is about to act on is the thing they just
      // touched.
      ctx.actions.pick(key);
    },
  },
    h("span", { class: "twist" }, open ? "▾" : "▸"),
    label,
    note ? h("span", { class: "muted note" }, note) : null,
  );

  // Controls sit on the row they belong to, at the right-hand end, and outside
  // the fold's contents. Outside, because a folder can be acted on without being
  // opened first — deleting one should not require expanding it — and because in
  // an open file they would otherwise be six hundred lines further down, past the
  // thing they act on.
  //
  // Beside the fold rather than inside it: the whole row is already a button, and
  // a button within a button is both invalid and a target nobody hits reliably.
  // So the two are siblings in a flex row, which is what puts them on one line
  // without either swallowing the other's clicks.
  const acts = controls ? controls() : null;
  const row = acts ? h("div", { class: "fold-row" }, head, acts) : head;

  if (!open) return row;
  return h("div", { class: "folded" }, row,
    // `tree`, because this one holds deeper folds rather than content. Each fold
    // carries its own indent, so the inset every other fold body takes would
    // compound once per level here and push a filename five levels down off the
    // side of a phone.
    h("div", { class: "inner tree" }, ...children()));
}

// crumbs is the header line: what is being read, and how much of it there is.
export function libraryHeader(state, kind) {
  const lib = state.library;
  if (!lib) return null;
  const files = (lib.files || []).filter((f) => (kind === "docs" ? isDocument(f) : !isDocument(f)));
  const lines = files.reduce((n, f) => n + (f.lines || 0), 0);
  return h("p", { class: "muted" },
    `${ellipsis(lib.root || "repository", 32)} · ${plural(files.length, "file")} · ${plural(lines, "line")}`);
}
